### ISSUE TITLE
Constructor generated by dataclasses fail to bind type vars on lambda statements in function fields.

### ISSUE THREAD

**Issue by mrenow** [microsoft/pyright#7326]
Hi! Thanks for Pyright, its awesome!

**Describe the bug**
This appears to be an issue particular to the constructor generated by @dataclasstransform. Function valued fields behave differently to normal when passing type information to lambda statements: for some reason the parameter of the lambda statement fails to resolve type variables. Below I give four examples, which I believe should all exhibit the same behaviour.

**Code or Screenshots**

```python
from typing import TypeVar, Callable
import dataclasses
T = TypeVar('T')


def generic_identity(val: T) -> T:
    return val


def identity(val: float) -> float:
    return val


@dataclasses.dataclass
class DataclassFunc:
    x: Callable[[float], float]


class ClassFunc:
    def __init__(self, x: Callable[[float], float]):
        self.x = x


def func_func(x: Callable[[float], float]):
    pass


ClassFunc(lambda r: generic_identity(r) + 1)  # OK
func_func(lambda r: generic_identity(r) + 1)  # OK
DataclassFunc(lambda r: generic_identity(r) + 1)  # information: Return type of lambda is unknown (reportUnknownLambdaType)
DataclassFunc(lambda r: identity(r) + 1)  # OK
```

**VS Code extension or command-line**
Tested on Pyright 1.1.351

**Comment by mrenow** [comment]
There is also strange behaviour when we throw overloads into the mix. I have not really figured this one out, but I believe it is related.

```python
from typing import TypeVar, Callable, overload, Never
import dataclasses
T = TypeVar('T')


@dataclasses.dataclass
class DataclassFunc:
    x: Callable[[float | None], float]


class ClassFunc:
    def __init__(self, x: Callable[[float | None], float]):
        self.x = x


def func_func(x: Callable[[float | None], float]):
    pass


@overload
def guard(val: None) -> Never: ...
@overload
def guard(val: T) -> T: ...


def guard(val: T | None) -> T:
    assert val is not None
    return val


ClassFunc(lambda r: guard(r))  # OK
func_func(lambda r: guard(r))  # OK
DataclassFunc(lambda r: guard(r))
'''
/home/ezra/Programming/Git/Bright/health/$loc$test.py
  /home/ezra/Programming/Git/Bright/health/$loc$test.py:32:31 - error: Argument of type "float | None" cannot be assigned to parameter "val" of type "None" in function "guard"
    Type "float | None" cannot be assigned to type "None"
      "float" is incompatible with "None" (reportArgumentType)
1 error, 0 warnings, 0 informations 
'''
```

**Comment by erictraut** [comment]
The problem is related to the `__new__` method that pyright synthesizes for a dataclass. If you change `ClassFunc` to the following, you'll see similar behavior:

```python
class ClassFunc:
    def __new__(cls, *args: Any, **kwargs: Any) -> "ClassFunc":
        return super().__new__(*args, **kwargs)

    def __init__(self, x: Callable[[float], float]) -> None:
        self.x = x
```

I'll continue to investigate, but I wanted to share what I've found so far.

**Comment by erictraut** [comment]
Found and fixed the underlying cause. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.352, which I just published.
